README.md
# jQuery Password Strength Meter for Twitter Bootstrap

[![Build Status](https://travis-ci.org/ablanco/jquery.pwstrength.bootstrap.png?branch=master)](https://travis-ci.org/ablanco/jquery.pwstrength.bootstrap)
[![Code Climate](https://codeclimate.com/github/ablanco/jquery.pwstrength.bootstrap.png)](https://codeclimate.com/github/ablanco/jquery.pwstrength.bootstrap)
[![devDependency Status](https://david-dm.org/ablanco/jquery.pwstrength.bootstrap/dev-status.png)](https://david-dm.org/ablanco/jquery.pwstrength.bootstrap#info=devDependencies)
[![](https://data.jsdelivr.com/v1/package/npm/pwstrength-bootstrap/badge?style=rounded)](https://www.jsdelivr.com/package/npm/pwstrength-bootstrap)

The jQuery Password Strength Meter is a plugin for Twitter Bootstrap that
provides rulesets for visualy displaying the quality of a users typed in
password.

Dual licensed under the MIT and GPL licenses. You can choose the one that
suits your purposes better.

[npm entry](https://www.npmjs.com/package/pwstrength-bootstrap)


## Requirements

* jQuery 1.7 or higher
* Bootstrap 2, 3 or 4

### Not using Bootstrap?

This plugin currently relies heavily on Bootstrap and it is not possible to
use it with another framework without making big changes in the code or
forgetting completely about the UI feedback.

Forks to use it with another frameworks that I know of:

* [Zurb Foundation fork by edtownend](https://github.com/edtownend/jquery.pwstrength.foundation)


## How to use it

Get the latest version through [Bower](http://bower.io/search/?q=pwstrength-bootstrap),
[npm](https://www.npmjs.com/package/pwstrength-bootstrap), or just download it
from this [repository](https://github.com/ablanco/jquery.pwstrength.bootstrap/tree/master/dist).
Load it into your HTML after your original bootstrap and jQuery javascript files:

```html
<script type="text/javascript" src="dist/pwstrength-bootstrap.min.js"></script>
```

Then just invoke the plugin on the password fields you want to attach a strength
meter to. For example, to use it on all the password fields with the default
examples:

```javascript
$(':password').pwstrength();
```

To apply it only to one input and change the options:

```javascript
$('#passwd1').pwstrength({
    ui: { showVerdictsInsideProgressBar: true }
});
```

## Options

Click here to find [the complete list of options for the plugin](OPTIONS.md).

If you are looking for options to change or add new texts, please have a look
at the internationalization section.


## Methods

Once the plugin has been initialized, it is possible to interact with it
through the methods.


### Force an update

It is possible to force an update on a password strength meter. It will force
a new score calculation and an update of the UI elements, the `onKeyUp`
callback will be called.

```javascript
$("#passwdfield").pwstrength("forceUpdate");
```


### Remove the strength meter

This will remove the data associated to the meter, and the UI elements.

```javascript
$("#passwdfield").pwstrength("destroy");
```


### Adding Custom Rules

The plugin comes with the functionality to easily define your own custom rules.
The format is as follows:

```javascript
$("#passwdfield").pwstrength("addRule", "ruleName", function (options, word, score) {}, rule_score, rule_enabled);
```

Example:

```javascript
$("#passwdfield").pwstrength("addRule", "testRule", function (options, word, score) {
    return word.match(/[a-z].[0-9]/) && score;
}, 10, true);
```


### Change the score associated to a rule

It is possible to change the score given by a rule. It works like this:

```javascript
$("#passwdfield").pwstrength("changeScore", "wordSequences", -100);
```

That would penalize even more the presence of sequences in the password.


### Activate and deactivate rules

It is also possible to activate or deactivate rules. It as simple as:

```javascript
$("#passwdfield").pwstrength("ruleActive", "wordSequences", false);
```

That would avoid looking for sequences in the password being tested.


### Know if all password inputs pass a specific rule

This method allows to make a quick check to test if all password inputs in the
page pass a rule, the method returns a boolean value. Example:

```javascript
$("#passwdfield").pwstrength("ruleIsMet", "wordSequences");
```


## Callback Functions

The plugin provides three callback functions, onLoad, onKeyUp, and scoreCalculated.  You can use
them like this:

```javascript
$(document).ready(function () {
    var options = {};
    options.common = {
        onLoad: function () {
            $('#messages').text('Start typing password');
        },
        onKeyUp: function (evt, data) {
            $("#length-help-text").text("Current length: " + $(evt.target).val().length + " and score: " + data.score);
        },
        onScore: function (options, word, totalScoreCalculated) {
            // If my word meets a specific scenario, I want the min score to
            // be the level 1 score, for example.
            if (word.length === 20 && totalScoreCalculated < options.ui.scores[1]) {
                // Score doesn't meet the score[1]. So we will return the min
                // numbers of points to get that score instead.
                return options.ui.score[1]
            }
            // Fall back to the score that was calculated by the rules engine.
            // Must pass back the score to set the total score variable.
            return totalScoreCalculated;
        }
    };
    $(':password').pwstrength(options);
});
```


## Extra restrictions

The plugin comes with two validation rules deactivated by default. One checks
the length of the password and penalizes it if it's too long; and the other
checks if the password contains a banned char, and penalizes it if it does.

You can configure the max length of the password by using the option `maxChar`.
You can also configure the invalid chars by using the option
`invalidCharsRegExp`.

If you need these restrictions you just need to activate this two rules:

```javascript
$(document).ready(function () {
    var options = {};
    options.rules = {
        activated: {
            wordMaxLength: true,
            wordInvalidChar: true
        }
    };
    $(':password').pwstrength(options);
});
```


## Internationalization (i18n)

The plugin has support for internationalization. It also comes with some
example translations, you can find them in the [locales folder](locales).

The plugin provides a default implementation of the translation function, but
you can override it using the option `i18n.t`.

The default implementation will try to make use of the popular
[i18next front-end translation tool](http://i18next.com/). If you happen to
use it, then you only need to add the translations into your resources and
load them. The plugin will automatically make use of it. You can find more
details about and how to use it i18next in their website. There is also an
example in the repository that uses that library.

In case the i18next library is not available, then the default behavior is
to return the english texts as a fallback.

### What are the translatable texts?

You can find the non-rules texts in any of the
[provided translation example files](locales), and besides what you find there,
every rule name is a valid key for the translation file. __You can use this to
add new error messages (or remove them) for the engine rules__.

### How to customize the translation function

If you want to manage translations yourself or you don't use i18next you can
override the default translation function like this:

```javascript
$(document).ready(function () {
    var options = {};
    options.i18n = {
        t: function (key) {
            var result = translateThisThing(key); // Do your magic here

            return result === key ? '' : result; // This assumes you return the
            // key if no translation was found, adapt as necessary
        }
    };
    $(':password').pwstrength(options);
});
```

You can find an example of some keys and translations in the
[locales folder](locales).


## Examples

There are some examples in the `examples` directory. Just serve them with any
webserver and check them in your browser. Make sure you serve the `examples`
directory as the site root. For example:

```bash
cd examples
python -m SimpleHTTPServer
```

And go to [localhost:8000](http://localhost:8000).

Alternatively, you can check-out the examples in a [hosted demo](https://cdn.rawgit.com/ablanco/jquery.pwstrength.bootstrap/master/examples/index.html).

## Build and Test

The build and testing processes rely on [Grunt](http://gruntjs.com/). To use
them you need to have [node.js](http://nodejs.org/) and grunt-cli installed on
your system. Assuming you have node.js in your Linux system, you'll need to do
something like this:

```bash
sudo npm install -g grunt-cli
```

Now you have the grunt command line utility installed globally.


### Bundle and minified

To generate the bundle and the minified file you only need to execute this in
the project directory:

```bash
npm install -d
grunt
```

It will check the source files, and build a minified version with its
corresponding source map. The generated files will be available in the `dist`
directory.


### Testing

To run the tests the only thing you need to do is execute this in the project
directory:

```bash
npm install -d
grunt test
```

It will check all the source files with [ESLint](https://eslint.org/) and run the
tests, which are written with [Jasmine](http://jasmine.github.io/). You'll find
the tests source code in the `spec` directory.

[Travis](https://travis-ci.org/ablanco/jquery.pwstrength.bootstrap) is being
used for continuos integration. You can check there if the tests are passing.


CHANGES.md
# CHANGELOG

## 3.1.0dev

- Allow chaining after setter methods.
- Bootstrap 5 support.

## 3.0.10

- Update dev dependencies.
- Stop using deprecated $.isFunction.

## 3.0.9

- Add jsdelivr entry in package.json.
- Update dev dependencies.

## 3.0.8

- Start using peerDependencies instead of regular dependencies.

## 3.0.7

- Fix dependencies version mess.

## 3.0.6

- Update dependencies.

## 3.0.5

- Apply Prettier code formatter.
- Replace JSLint with ESLint.
- Add new rules.specialCharClass option to personalize special chars
  detection.
- Separate progress bar and popover widgets to their own files.

## 3.0.4

- Fix ruleIsMet method, include added custom validation rules.
- Fix conflict between progressBarMinWidth and progressBarEmptyPercentage
  options.

## 3.0.3

- Add Arabic localization.
- Add Norwegian localization.
- Add new ui.progressBarMinWidth option to control the minimum width of the
  progress bar.

## 3.0.2

- Add Czech localization.
- Bugfix with popovers in Bootstrap 4.2.X versions.

## 3.0.1

- Bugfix in progress bar colors in Bootstrap 3.

## 3.0.0

- Bootstrap 4 by default.
- Improved support for Bootstrap 4.

## 2.2.1

- Bugfix in the common passwords rule.

## 2.2.0

- Add new rule to penalize common passwords.

## 2.1.4

- Thai localization.
- Fix typo in German localization.
- Activate by default the extra security rules.
- Make the invalid chars optional rule configurable.

## 2.1.3

- Bugfix, call `onScore` when zxcvbn is in use too.

## 2.1.2

- Fix errors in Portuguese localization.
- Fix French localization capitalization.
- Fix ruleIsMet issues with wordMin and wordMax rules.
- Don't allow verdict to break line when inside progress bar.

## 2.1.1

- Add missing rule, needed by the `ruleIsMet` method.
- Add `wordMaxLength` and `wordInvalidChar` optional rules to the engine.

## 2.1.0

- Slovak translation.
- Add a new `ruleIsMet` method that returns a boolean value indicating if all
  password inputs in the page pass a specific rule.

## 2.0.8

- Fix showing the strength of the password through the status of the field.

## 2.0.7

- Add new option `progressExtraCssClasses` to be able to customize the
  container of the progress bar.
- Updated development dependencies.

## 2.0.6

- Updated development dependencies.
- Bootstrap 4 alpha 6 support.

## 2.0.5

- Italian localization.

## 2.0.4

- French localization.
- Don't use Math.log2 since IE doesn't support it.

## 2.0.3

- German localization.
- Polish localization.

## 2.0.2

- Add a `onScore` callback to allow for a final score modification.
- Turkish localization.

## 2.0.1

- Fix bad assignment in the plugin initialization.
- Russian localization.
- New option to control the events the plugin listen to.

## 2.0.0

- Use six possible verdicts and six possible css classes, so they match one
  to one making it possible to configure each class for each verdict level.
- Properly manage the paste event so the meter updates when the user pastes the
  password.
- Add a new option to display the password score.
- Translations support, ahora hablamos idiomas.
- New option to set the minimum possible percentage filled in the progress bar
  when the password field is not empty.
- New option to set the minimum possible percentage filled in the progress bar
  when the password field is empty.
- New option for extra CSS classes to be added to the generated progress bar.

### Breaking changes

- There are 6 verdicts and css classes now, instead of 5.
- `verdicts` and `errorMessages` options have been removed. Now they rely on
  the translations system.

## 1.2.10

- Replace entropy call with log2 of guesses for zxcvbn because entropy property
  is removed in zxcvbn v4.0.1, and it was just log2 of guesses.

## 1.2.9

- No changes, I forgot to add the built files into the 1.2.8, so I'm releasing
  the same again.

## 1.2.8

- Updated to work with Bootstrap 4. Bootstrap 3 is still the default mode.
- Allow to establish the placement of the popover through an option.
- Make the css classes added to the bar and verdicts customizable.
- Bugfix in the progress bar percentage calculation for a score of zero.

## 1.2.7

- Bugfix: escape special characters in username for regex.

## 1.2.6

- More sensible default score for sequences rule.
- Publish plugin in npm.

## 1.2.5

- Bugfix when using zxcvbn and form inputs with empty values.
- New option to specify a list of banned words for zxcvbn.

## 1.2.4

- New option to add a class in verdict element.
- If there is text in the password field, don't show the progress bar empty.
- Empty verdict for an empty password field.
- Support html in the verdicts content.

## 1.2.3

- New option to customize the html of the popover with the errors.
- Bugfix in special char regex.

## 1.2.2

- Every rule can have associated error messages.

## 1.2.1

- Improve documentation.
- Fix typo in alphabetical sequence.
- Use the not minified version of the library in bower as main file.

## 1.2.0

- Listen also to the `change` and `onpaste` events, not only to the `onkeyup`.
- Show the lowest verdict when the score is below zero.
- New option to pass more input fields content to the zxcvbn library.
- Don't show the verdicts inside the popover if they are being showed inside
  the progressbar.

## 1.1.5

- Better Bower configuration.
- Pass also the verdict level to the "on key up" event handler.
- Add a basic usage section to the readme.

## 1.1.4

- Bower support.

## 1.1.3

- Pass the score and the verdict to the "on key up" event handler.

## 1.1.2

- Upgrade dev dependencies: grunt plugins and jquery
- Bugfix in sequences lookup
- New tests for sequences lookup

## 1.1.1

- Pass the username field content to the zxcvbn function, so zxcvbn takes it
  into consideration when scoring the password.
- Add a debug option, so the score gets printed in the JS console.
- Check reversed sequences too in the sequences rule.
- Fix the popover flickering.

## 1.1.0


spec/rules.spec.js
/* eslint-env node, jasmine */

'use strict';

var jsdom = require('jsdom').jsdom,
    jQuery = require('jquery'),
    rulesEngine = require('../src/rules.js'),
    options = {
        common: {
            minChar: 6,
            usernameField: '#username'
        },
        rules: {
            activated: {
                wordNotEmail: true,
                wordMinLength: true,
                wordMaxLength: false,
                wordInvalidChar: false,
                wordSimilarToUsername: true,
                wordSequences: true,
                wordTwoCharacterClasses: false,
                wordRepetitions: false,
                wordLowercase: true,
                wordUppercase: true,
                wordOneNumber: true,
                wordThreeNumbers: true,
                wordOneSpecialChar: true,
                wordTwoSpecialChar: true,
                wordUpperLowerCombo: true,
                wordLetterNumberCombo: true,
                wordLetterNumberCharCombo: true,
                wordIsACommonPassword: true
            },
            scores: {
                wordNotEmail: -100,
                wordMinLength: -50,
                wordMaxLength: -50,
                wordInvalidChar: -100,
                wordSimilarToUsername: -100,
                wordSequences: -50,
                wordTwoCharacterClasses: 2,
                wordRepetitions: -25,
                wordLowercase: 1,
                wordUppercase: 3,
                wordOneNumber: 3,
                wordThreeNumbers: 5,
                wordOneSpecialChar: 3,
                wordTwoSpecialChar: 5,
                wordUpperLowerCombo: 2,
                wordLetterNumberCombo: 2,
                wordLetterNumberCharCombo: 2,
                wordIsACommonPassword: -100
            },
            raisePower: 1.4,
            commonPasswords: ['123456', 'password', '12345678', 'qwerty']
        },
        ui: {
            spanError: function(ignore, rule) {
                return rule;
            }
        },
        instances: {
            errors: []
        }
    },
    strictOptions,
    relaxedOptions,
    $;

$ = jQuery(jsdom().defaultView);

strictOptions = $.extend(true, {}, options);
strictOptions.rules.activated.wordTwoCharacterClasses = true;
strictOptions.rules.activated.wordRepetitions = true;

relaxedOptions = $.extend(true, {}, options);
relaxedOptions.rules.activated.wordNotEmail = false;
relaxedOptions.rules.activated.wordMinLength = false;
relaxedOptions.rules.activated.wordSequences = false;

describe('Applying default rules', function() {
    it('to a common "password"', function() {
        var password = 'password';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            -80.62082632004744
        );
    });

    it('to a mixed "charsand19283746"', function() {
        var password = 'charsand19283746';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            59.50293012833273
        );
    });

    it('to one with upper cases', function() {
        var password = 'QAZwsxTGByhn';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            38.423040924494714
        );
    });

    it('to one with symbols', function() {
        var password = 'qpwo#(ei*$&%^';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            47.26775666655811
        );
    });

    it('to one with everything', function() {
        var password = '3vEr!t#iNg1$f!n3';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            74.50293012833274
        );
    });

    it('to the perfect one', function() {
        var password = 'qm2oUY!%$32znheSK&*3@#';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            101.75159003283396
        );
    });

    it('to one with sequences', function() {
        var password = 'qwertypasswdisbad';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            3.799339527161436
        );
    });

    it('to one with reversed sequences', function() {
        var password = '0987ytrewqisbad';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            5.312654085941652
        );
    });

    it('to one with an email', function() {
        var password = 'test@example.com';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            -45.49706987166727
        );
    });

    it('to a "short" one', function() {
        var password = 'short';
        expect(rulesEngine.executeRules(options, password)).toEqual(
            -39.48173030642061
        );
    });
});

describe('Applying stricter rules', function() {
    it('to two-classes "passWORD"', function() {
        var password = 'passWORD';
        expect(rulesEngine.executeRules(strictOptions, password)).toEqual(
            26.379173679952558
        );
    });

    it('to one with repetitions', function() {
        var password = 'rreepppeatttt';
        expect(rulesEngine.executeRules(strictOptions, password)).toEqual(
            12.267756666558107
        );
    });
});

describe('Applying relaxed rules', function() {
    it('to one with sequences', function() {
        var password = 'qwertypasswdisbad';
        expect(rulesEngine.executeRules(relaxedOptions, password)).toEqual(1);
    });

    it('to one with an email', function() {
        var password = 'test@example.com';
        expect(rulesEngine.executeRules(relaxedOptions, password)).toEqual(6);
    });

    it('to a "short" one', function() {
        var password = 'short';
        expect(rulesEngine.executeRules(relaxedOptions, password)).toEqual(1);
    });
});

describe('Looking for sequences', function() {
    beforeEach(function() {
        options.instances.errors = [];
    });

    it('at the beginning of the password', function() {
        var password = 'abcx';
        expect(
            rulesEngine.validation.wordSequences(options, password, 123)
        ).toEqual(123);
    });

    it('at the end of the password', function() {
        var password = 'xabc';
        expect(
            rulesEngine.validation.wordSequences(options, password, 123)
        ).toEqual(123);
    });

    it('and not finding them', function() {
        var password = 'mdiw93jc65oak$&!';
        expect(
            rulesEngine.validation.wordSequences(options, password, 123)
        ).toEqual(0);
    });
});


src/ui.js
/*
 * jQuery Password Strength plugin for Twitter Bootstrap
 *
 * Copyright (c) 2008-2013 Tane Piper
 * Copyright (c) 2013 Alejandro Blanco
 * Dual licensed under the MIT and GPL licenses.
 */

// eslint-disable-next-line no-implicit-globals
var ui = {};

(function($) {
    'use strict';

    var statusClasses = ['error', 'warning', 'success'],
        verdictKeys = [
            'veryWeak',
            'weak',
            'normal',
            'medium',
            'strong',
            'veryStrong'
        ];

    ui.getContainer = function(options, $el) {
        var $container;

        $container = $(options.ui.container);
        if (!($container && $container.length === 1)) {
            $container = $el.parent();
        }
        return $container;
    };

    ui.findElement = function($container, viewport, cssSelector) {
        if (viewport) {
            return $container.find(viewport).find(cssSelector);
        }
        return $container.find(cssSelector);
    };

    ui.getUIElements = function(options, $el) {
        var $container, result;

        if (options.instances.viewports) {
            return options.instances.viewports;
        }

        $container = ui.getContainer(options, $el);

        result = {};
        result.$progressbar = ui.findElement(
            $container,
            options.ui.viewports.progress,
            'div.progress'
        );
        if (options.ui.showVerdictsInsideProgressBar) {
            result.$verdict = result.$progressbar.find('span.password-verdict');
        }

        if (!options.ui.showPopover) {
            if (!options.ui.showVerdictsInsideProgressBar) {
                result.$verdict = ui.findElement(
                    $container,
                    options.ui.viewports.verdict,
                    'span.password-verdict'
                );
            }
            result.$errors = ui.findElement(
                $container,
                options.ui.viewports.errors,
                'ul.error-list'
            );
        }
        result.$score = ui.findElement(
            $container,
            options.ui.viewports.score,
            'span.password-score'
        );

        options.instances.viewports = result;
        return result;
    };

    ui.initHelper = function(options, $el, html, viewport) {
        var $container = ui.getContainer(options, $el);
        if (viewport) {
            $container.find(viewport).append(html);
        } else {
            $(html).insertAfter($el);
        }
    };

    ui.initVerdict = function(options, $el) {
        ui.initHelper(
            options,
            $el,
            '<span class="password-verdict"></span>',
            options.ui.viewports.verdict
        );
    };

    ui.initErrorList = function(options, $el) {
        ui.initHelper(
            options,
            $el,
            '<ul class="error-list"></ul>',
            options.ui.viewports.errors
        );
    };

    ui.initScore = function(options, $el) {
        ui.initHelper(
            options,
            $el,
            '<span class="password-score"></span>',
            options.ui.viewports.score
        );
    };

    ui.initUI = function(options, $el) {
        if (options.ui.showPopover) {
            ui.initPopover(options, $el);
        } else {
            if (options.ui.showErrors) {
                ui.initErrorList(options, $el);
            }
            if (
                options.ui.showVerdicts &&
                !options.ui.showVerdictsInsideProgressBar
            ) {
                ui.initVerdict(options, $el);
            }
        }
        if (options.ui.showProgressBar) {
            ui.initProgressBar(options, $el);
        }
        if (options.ui.showScore) {
            ui.initScore(options, $el);
        }
    };

    ui.updateVerdict = function(options, $el, cssClass, text) {
        var $verdict = ui.getUIElements(options, $el).$verdict;
        $verdict.removeClass(options.ui.colorClasses.join(' '));
        if (cssClass > -1) {
            $verdict.addClass(options.ui.colorClasses[cssClass]);
        }
        if (options.ui.showVerdictsInsideProgressBar) {
            $verdict.css('white-space', 'nowrap');
        }
        $verdict.html(text);
    };

    ui.updateErrors = function(options, $el, remove) {
        var $errors = ui.getUIElements(options, $el).$errors,
            html = '';

        if (!remove) {
            $.each(options.instances.errors, function(idx, err) {
                html += '<li>' + err + '</li>';
            });
        }
        $errors.html(html);
    };

    ui.updateScore = function(options, $el, score, remove) {
        var $score = ui.getUIElements(options, $el).$score,
            html = '';

        if (!remove) {
            html = score.toFixed(2);
        }
        $score.html(html);
    };

    ui.updateFieldStatus = function(options, $el, cssClass, remove) {
        var $target = $el;

        if (options.ui.bootstrap2) {
            $target = $el.parents('.control-group').first();
        } else if (options.ui.bootstrap3) {
            $target = $el.parents('.form-group').first();
        }

        $.each(statusClasses, function(idx, css) {
            if (options.ui.bootstrap3) {
                css = 'has-' + css;
            } else if (!options.ui.bootstrap2) {
                // BS4
                if (css === 'error') {
                    css = 'danger';
                }
                css = 'border-' + css;
            }
            $target.removeClass(css);
        });

        if (remove) {
            return;
        }

        cssClass = statusClasses[Math.floor(cssClass / 2)];
        if (options.ui.bootstrap3) {
            cssClass = 'has-' + cssClass;
        } else if (!options.ui.bootstrap2) {
            // BS4
            if (cssClass === 'error') {
                cssClass = 'danger';
            }
            cssClass = 'border-' + cssClass;
        }
        $target.addClass(cssClass);
    };

    ui.getVerdictAndCssClass = function(options, score) {
        var level, verdict;

        if (score === undefined) {
            return ['', 0];
        }

        if (score <= options.ui.scores[0]) {
            level = 0;
        } else if (score < options.ui.scores[1]) {
            level = 1;
        } else if (score < options.ui.scores[2]) {
            level = 2;
        } else if (score < options.ui.scores[3]) {
            level = 3;
        } else if (score < options.ui.scores[4]) {
            level = 4;
        } else {
            level = 5;
        }

        verdict = verdictKeys[level];

        return [options.i18n.t(verdict), level];
    };

    ui.updateUI = function(options, $el, score) {
        var cssClass, barPercentage, verdictText, verdictCssClass;

        cssClass = ui.getVerdictAndCssClass(options, score);
        verdictText = score === 0 ? '' : cssClass[0];
        cssClass = cssClass[1];
        verdictCssClass = options.ui.useVerdictCssClass ? cssClass : -1;

        if (options.ui.showProgressBar) {
            if (score === undefined) {
                barPercentage = options.ui.progressBarEmptyPercentage;
            } else {
                barPercentage = ui.percentage(
                    options,
                    score,
                    options.ui.scores[4]
                );
            }
            ui.updateProgressBar(options, $el, cssClass, barPercentage);
            if (options.ui.showVerdictsInsideProgressBar) {
                ui.updateVerdict(options, $el, verdictCssClass, verdictText);
            }
        }

        if (options.ui.showStatus) {
            ui.updateFieldStatus(options, $el, cssClass, score === undefined);
        }

        if (options.ui.showPopover) {
            ui.updatePopover(options, $el, verdictText, score === undefined);
        } else {
            if (
                options.ui.showVerdicts &&
                !options.ui.showVerdictsInsideProgressBar
            ) {
                ui.updateVerdict(options, $el, verdictCssClass, verdictText);
            }
            if (options.ui.showErrors) {
                ui.updateErrors(options, $el, score === undefined);
            }
        }

        if (options.ui.showScore) {
            ui.updateScore(options, $el, score, score === undefined);
        }
    };
})(jQuery);


src/options.js
/*global i18n */

/*
 * jQuery Password Strength plugin for Twitter Bootstrap
 *
 * Copyright (c) 2008-2013 Tane Piper
 * Copyright (c) 2013 Alejandro Blanco
 * Dual licensed under the MIT and GPL licenses.
 */

// eslint-disable-next-line no-implicit-globals
var defaultOptions = {};

defaultOptions.common = {};
defaultOptions.common.minChar = 6;
defaultOptions.common.maxChar = 20;
defaultOptions.common.usernameField = '#username';
defaultOptions.common.invalidCharsRegExp = new RegExp(/[\s,'"]/);
defaultOptions.common.userInputs = [
    // Selectors for input fields with user input
];
defaultOptions.common.onLoad = undefined;
defaultOptions.common.onKeyUp = undefined;
defaultOptions.common.onScore = undefined;
defaultOptions.common.zxcvbn = false;
defaultOptions.common.zxcvbnTerms = [
    // List of disrecommended words
];
defaultOptions.common.events = ['keyup', 'change', 'paste'];
defaultOptions.common.debug = false;

defaultOptions.rules = {};
defaultOptions.rules.extra = {};
defaultOptions.rules.scores = {
    wordNotEmail: -100,
    wordMinLength: -50,
    wordMaxLength: -50,
    wordInvalidChar: -100,
    wordSimilarToUsername: -100,
    wordSequences: -20,
    wordTwoCharacterClasses: 2,
    wordRepetitions: -25,
    wordLowercase: 1,
    wordUppercase: 3,
    wordOneNumber: 3,
    wordThreeNumbers: 5,
    wordOneSpecialChar: 3,
    wordTwoSpecialChar: 5,
    wordUpperLowerCombo: 2,
    wordLetterNumberCombo: 2,
    wordLetterNumberCharCombo: 2,
    wordIsACommonPassword: -100
};
defaultOptions.rules.activated = {
    wordNotEmail: true,
    wordMinLength: true,
    wordMaxLength: false,
    wordInvalidChar: false,
    wordSimilarToUsername: true,
    wordSequences: true,
    wordTwoCharacterClasses: true,
    wordRepetitions: true,
    wordLowercase: true,
    wordUppercase: true,
    wordOneNumber: true,
    wordThreeNumbers: true,
    wordOneSpecialChar: true,
    wordTwoSpecialChar: true,
    wordUpperLowerCombo: true,
    wordLetterNumberCombo: true,
    wordLetterNumberCharCombo: true,
    wordIsACommonPassword: true
};
defaultOptions.rules.raisePower = 1.4;
// List taken from https://github.com/danielmiessler/SecLists (MIT License)
defaultOptions.rules.commonPasswords = [
    '123456',
    'password',
    '12345678',
    'qwerty',
    '123456789',
    '12345',
    '1234',
    '111111',
    '1234567',
    'dragon',
    '123123',
    'baseball',
    'abc123',
    'football',
    'monkey',
    'letmein',
    '696969',
    'shadow',
    'master',
    '666666',
    'qwertyuiop',
    '123321',
    'mustang',
    '1234567890',
    'michael',
    '654321',
    'pussy',
    'superman',
    '1qaz2wsx',
    '7777777',
    'fuckyou',
    '121212',
    '000000',
    'qazwsx',
    '123qwe',
    'killer',
    'trustno1',
    'jordan',
    'jennifer',
    'zxcvbnm',
    'asdfgh',
    'hunter',
    'buster',
    'soccer',
    'harley',
    'batman',
    'andrew',
    'tigger',
    'sunshine',
    'iloveyou',
    'fuckme',
    '2000',
    'charlie',
    'robert',
    'thomas',
    'hockey',
    'ranger',
    'daniel',
    'starwars',
    'klaster',
    '112233',
    'george',
    'asshole',
    'computer',
    'michelle',
    'jessica',
    'pepper',
    '1111',
    'zxcvbn',
    '555555',
    '11111111',
    '131313',
    'freedom',
    '777777',
    'pass',
    'fuck',
    'maggie',
    '159753',
    'aaaaaa',
    'ginger',
    'princess',
    'joshua',
    'cheese',
    'amanda',
    'summer',
    'love',
    'ashley',
    '6969',
    'nicole',
    'chelsea',
    'biteme',
    'matthew',
    'access',
    'yankees',
    '987654321',
    'dallas',
    'austin',
    'thunder',
    'taylor',
    'matrix'
];

defaultOptions.ui = {};
defaultOptions.ui.bootstrap2 = false;
defaultOptions.ui.bootstrap3 = false;
defaultOptions.ui.colorClasses = [
    'danger',
    'danger',
    'danger',
    'warning',
    'warning',
    'success'
];
defaultOptions.ui.showProgressBar = true;
defaultOptions.ui.progressBarEmptyPercentage = 1;
defaultOptions.ui.progressBarMinWidth = 1;
defaultOptions.ui.progressBarMinPercentage = 1;
defaultOptions.ui.progressExtraCssClasses = '';
defaultOptions.ui.progressBarExtraCssClasses = '';
defaultOptions.ui.showPopover = false;
defaultOptions.ui.popoverPlacement = 'bottom';
defaultOptions.ui.showStatus = false;
defaultOptions.ui.spanError = function(options, key) {
    'use strict';
    var text = options.i18n.t(key);
    if (!text) {
        return '';
    }
    return '<span style="color: #d52929">' + text + '</span>';
};
defaultOptions.ui.popoverError = function(options) {
    'use strict';
    var errors = options.instances.errors,
        errorsTitle = options.i18n.t('errorList'),
        message =
            '<div>' +
            errorsTitle +
            '<ul class="error-list" style="margin-bottom: 0;">';

    jQuery.each(errors, function(idx, err) {
        message += '<li>' + err + '</li>';
    });
    message += '</ul></div>';
    return message;
};
defaultOptions.ui.showVerdicts = true;
defaultOptions.ui.showVerdictsInsideProgressBar = false;
defaultOptions.ui.useVerdictCssClass = false;
defaultOptions.ui.showErrors = false;
defaultOptions.ui.showScore = false;
defaultOptions.ui.container = undefined;
defaultOptions.ui.viewports = {
    progress: undefined,
    verdict: undefined,
    errors: undefined,
    score: undefined
};
defaultOptions.ui.scores = [0, 14, 26, 38, 50];

defaultOptions.i18n = {};
defaultOptions.i18n.t = i18n.t;
